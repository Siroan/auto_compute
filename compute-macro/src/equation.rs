use proc_macro::TokenStream;
use syn::{Data, Fields};

use crate::field::Field;
use crate::symbols::VARIABLE;

fn log_structure(message: String) {
    logger::log(logger::LogStep::Structure, &message);
}

pub fn expand_derive_equation(input: &mut syn::DeriveInput) -> TokenStream {
    let mut variables = vec![];

    log_structure(format!("Parsing equation {:?}", input.ident.to_string()));

    let ident = &input.ident;
    let mut struct_diagnostics = quote! {};
    if let Data::Struct(data) = input.clone().data {
        if let Fields::Named(fields) = data.fields {
            fields.named.iter().for_each(|field| {
                for attr in &field.attrs {
                    if attr.path() == VARIABLE {
                        if let Some(ident) = field.clone().ident {
                            log_structure(format!("Found variable: {:?}", ident.to_string()));
                            variables.push(Field { name: ident });
                        }
                    } else {
                        break;
                    }
                }
            });
        } else {
            struct_diagnostics = quote! {
                #struct_diagnostics
                compile_error!("The equation must contain named fields");
            }
        }
    } else {
        struct_diagnostics = quote! {
            #struct_diagnostics
            compile_error!("The equation must be a struct");
        }
    }

    let mut find_unknown = quote! {};
    for variable in variables {
        let name = variable.name;
        let name_s = name.to_string();
        find_unknown = quote! {
            #find_unknown

            log_setup(format!("Variable \"{}\" is {:?}", #name_s, self.#name));

            if self.#name.is_unknown() {
                if unknown.is_some() {
                    log_setup(format!("Error: Several unknown"));
                    return Err(Error::SeveralUnknown);
                }
                unknown = Some(self.#name.clone());
            }
        }
    }

    quote! {
        use compute::error::Error;

        #[automatically_derived]
        impl #ident {
            fn compute(&self) -> Result<f64, Error> {
                use compute::equation::EquationElement;
                use compute::error::Error;

                #struct_diagnostics

                fn log_setup(message: String) {
                    logger::log(logger::LogStep::Setup, &message);
                }

                let mut unknown = None;
                #find_unknown

                match unknown {
                    None => {
                        log_setup(format!("Error: No unknown"));
                        Err(Error::NoUnkown)
                    },
                    Some(unknown) => {
                        self.auto_compute();
                        unknown.get_unknown_value()
                    }
                }
            }
        }
    }
    .into()
}
