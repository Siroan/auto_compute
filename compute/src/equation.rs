use crate::unknown::Unknown;

pub trait EquationAutoCompute {
    fn auto_compute(&self) -> bool;
}

#[derive(Clone, Debug, PartialEq)]
pub enum EquationElement {
    Known(f64),
    Unknown(Unknown),
}

impl EquationElement {
    pub fn new_unknown() -> EquationElement {
        EquationElement::Unknown(Unknown::new())
    }
}
