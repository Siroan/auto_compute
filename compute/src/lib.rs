pub mod unknown;
pub mod element;
pub mod equation;
pub mod error;