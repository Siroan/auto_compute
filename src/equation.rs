extern crate proc_macro;
use proc_macro::TokenStream;

use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

#[derive(Debug)]
pub enum PodElement {
    Known(f64),
    Unknown,
    //UnknownWithF-1
}

pub type Pod = HashMap<String, PodElement>;

struct Equation {
    x: Rc<RefCell<f64>>,
    pod: Pod,
    compute: Box<dyn Fn() -> bool>,
}

impl Equation {
    fn new(pod: Pod, compute: Box<dyn Fn() -> bool>) -> Self {
        Self {
            x: Rc::new(RefCell::new(0.)),
            pod,
            compute,
        }
    }

    fn compute(&self) -> f64 {
        self.compute();
        *self.x.borrow()
    }
}

//#[proc_macro_derive(AnswerFn)]
//pub fn derive_answer_fn(_item: TokenStream) -> TokenStream {
//    "fn answer() -> u32 { 42 }".parse().unwrap()
//}
