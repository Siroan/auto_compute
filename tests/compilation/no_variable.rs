extern crate compute;

#[macro_use]
extern crate compute_macro;

use compute::equation::{EquationAutoCompute, EquationElement};

fn main() {
    #[derive(Equation)]
    struct NoVariable {
        pub member: EquationElement,
    }

    impl EquationAutoCompute for NoVariable {
        fn auto_compute(&self) -> bool {
            true
        }
    }
}