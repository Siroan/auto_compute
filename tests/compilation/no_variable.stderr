error: There is no variable, you can add the attribute #[variable] to a field to create one
 --> tests/compilation/no_variable.rs:9:14
  |
9 |     #[derive(Equation)]
  |              ^^^^^^^^
  |
  = note: this error originates in the derive macro `Equation` (in Nightly builds, run with -Z macro-backtrace for more info)
